File: mobile-test-agent/mobile_test_agent/verify.py
"""Post-action verification — the third leg of find → act → VERIFY.

Every command type has an explicit postcondition checked against a fresh
snapshot.  A step only PASSES when its postcondition holds; otherwise the
orchestrator retries with a recovery ladder.  This is what pushes the
end-to-end failure rate down: an action that silently did nothing is
caught immediately, on the step where it happened.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from .matcher import norm, resolve, text_score
from .parser import Command
from .snapshot import Snapshot, UIElement


@dataclass
class Verification:
    ok: bool
    detail: str = ""
    weak: bool = False   # action ran, postcondition unobservable (still usable)


def verify_tap(before: Snapshot, after: Snapshot, el: UIElement) -> Verification:
    if after.hash != before.hash:
        return Verification(True, "screen changed after tap")
    fresh = _find_same(after, el)
    if fresh is None:
        return Verification(True, "tapped element left the screen")
    if fresh.selected != el.selected or fresh.checked != el.checked:
        return Verification(True, "element state changed")
    return Verification(False, "no observable change after tap")


def verify_input(after: Snapshot, el: UIElement, cmd: Command) -> Verification:
    fresh = _find_same(after, el)
    if fresh is None:
        return Verification(True, "field not re-found (screen moved on)", weak=True)
    content = fresh.text or fresh.value
    if cmd.secure or fresh.password:
        # masked text ("•••••") normalises to "", so compare raw strings
        ok = bool(content.strip()) and content.strip() != (el.hint or "").strip()
        return Verification(ok, "secure field is non-empty" if ok
                            else "secure field still empty")
    if norm(content) == norm(cmd.value or ""):
        return Verification(True, "field text matches")
    if cmd.value and (cmd.value in (content or "")):
        return Verification(True, "field contains entered text")
    # masked/formatted inputs (e.g. auto-formatted dates) — accept non-empty
    if content and norm(content) != norm(el.hint or "") and content != el.text:
        return Verification(True, f"field changed to '{content[:30]}'", weak=True)
    return Verification(False, f"field shows '{(content or '')[:30]}', "
                               f"expected '{(cmd.value or '')[:30]}'")


def verify_visible(snap: Snapshot, cmd: Command, accept: float = 0.72) -> Verification:
    res = resolve(Command(kind="tap", target=cmd.target, raw=cmd.raw), snap,
                  accept=accept, margin=0.0, strong=accept)
    if cmd.kind == "assert_not_visible":
        ok = res.element is None and res.score < accept
        return Verification(ok, "element absent" if ok else "element still visible")
    ok = res.element is not None or res.score >= accept
    return Verification(ok, "element visible" if ok else
                        f"best match score {res.score:.2f} below {accept}")


def verify_date_shown(snap: Snapshot, el: Optional[UIElement],
                      value_variants: list) -> Verification:
    """After a picker flow, the field should now display the chosen date."""
    if el is None:
        return Verification(True, "no anchor field to re-check", weak=True)
    fresh = _find_same(snap, el)
    content = (fresh.text or fresh.value) if fresh else ""
    for v in value_variants:
        if v and text_score(v, content) >= 0.6:
            return Verification(True, f"field shows '{content[:30]}'")
    if content:
        return Verification(True, f"field changed to '{content[:30]}'", weak=True)
    return Verification(False, "date field still empty")


def _find_same(snap: Snapshot, old: UIElement) -> Optional[UIElement]:
    for el in snap.elements:
        if old.res_id and el.res_id == old.res_id:
            return el
    for el in snap.elements:
        if el.center == old.center and el.tag == old.tag:
            return el
    return None

File: mobile-test-agent/src/verify.ts
/**
 * Post-action verification — the third leg of find → act → VERIFY.
 *
 * Every command type has an explicit postcondition checked against a
 * fresh snapshot. A step only PASSES when its postcondition holds;
 * otherwise the orchestrator retries with a recovery ladder. This is
 * what pushes the end-to-end failure rate down: an action that silently
 * did nothing is caught immediately, on the step where it happened.
 */
import { norm, resolveTarget, textScore } from "./matcher.js";
import type { Command } from "./parser.js";
import { center, type Snapshot, type UIElement } from "./snapshot.js";

export interface Verification {
  ok: boolean;
  detail: string;
  weak?: boolean;  // action ran, postcondition unobservable (still usable)
}

export function verifyTap(
  before: Snapshot, after: Snapshot, el: UIElement,
): Verification {
  if (after.hash !== before.hash) {
    return { ok: true, detail: "screen changed after tap" };
  }
  const fresh = findSame(after, el);
  if (!fresh) return { ok: true, detail: "tapped element left the screen" };
  if (fresh.selected !== el.selected || fresh.checked !== el.checked) {
    return { ok: true, detail: "element state changed" };
  }
  return { ok: false, detail: "no observable change after tap" };
}

export function verifyInput(
  after: Snapshot, el: UIElement, command: Command,
): Verification {
  const fresh = findSame(after, el);
  if (!fresh) {
    return { ok: true, detail: "field not re-found (screen moved on)", weak: true };
  }
  const content = fresh.text || fresh.value;
  if (command.secure || fresh.password) {
    // masked text ("•••••") normalises to "", so compare raw strings
    const ok = !!content.trim() && content.trim() !== (el.hint ?? "").trim();
    return {
      ok,
      detail: ok ? "secure field is non-empty" : "secure field still empty",
    };
  }
  if (norm(content) === norm(command.value ?? "")) {
    return { ok: true, detail: "field text matches" };
  }
  if (command.value && content.includes(command.value)) {
    return { ok: true, detail: "field contains entered text" };
  }
  // masked/formatted inputs (e.g. auto-formatted dates) — accept non-empty
  if (content && norm(content) !== norm(el.hint ?? "") && content !== el.text) {
    return { ok: true, detail: `field changed to '${content.slice(0, 30)}'`, weak: true };
  }
  return {
    ok: false,
    detail: `field shows '${content.slice(0, 30)}', ` +
            `expected '${(command.value ?? "").slice(0, 30)}'`,
  };
}

export function verifyVisible(
  snap: Snapshot, command: Command, accept = 0.72,
): Verification {
  const probe: Command = {
    kind: "tap", target: command.target,
    secure: false, maybeDate: false, raw: command.raw,
  };
  const res = resolveTarget(probe, snap, accept, 0, accept);
  if (command.kind === "assert_not_visible") {
    const ok = !res.element && res.score < accept;
    return { ok, detail: ok ? "element absent" : "element still visible" };
  }
  const ok = !!res.element || res.score >= accept;
  return {
    ok,
    detail: ok ? "element visible"
      : `best match score ${res.score.toFixed(2)} below ${accept}`,
  };
}

/** After a picker flow, the field should now display the chosen date. */
export function verifyDateShown(
  snap: Snapshot, el: UIElement | undefined, valueVariants: string[],
): Verification {
  if (!el) return { ok: true, detail: "no anchor field to re-check", weak: true };
  const fresh = findSame(snap, el);
  const content = fresh ? fresh.text || fresh.value : "";
  for (const v of valueVariants) {
    if (v && textScore(v, content) >= 0.6) {
      return { ok: true, detail: `field shows '${content.slice(0, 30)}'` };
    }
  }
  if (content) {
    return { ok: true, detail: `field changed to '${content.slice(0, 30)}'`, weak: true };
  }
  return { ok: false, detail: "date field still empty" };
}

function findSame(snap: Snapshot, old: UIElement): UIElement | undefined {
  for (const el of snap.elements) {
    if (old.resId && el.resId === old.resId) return el;
  }
  const [ox, oy] = center(old);
  for (const el of snap.elements) {
    const [x, y] = center(el);
    if (x === ox && y === oy && el.tag === old.tag) return el;
  }
  return undefined;
}
